File: 2025/Day11/solution_part2.scala
// Advent of Code 2025 - Day 11: Reactor (Part 2)

import scala.io.Source
import scala.collection.mutable

object Solution {

  /** Parse input into adjacency list: device -> list of output devices */
  def parseGraph(lines: Seq[String]): Map[String, List[String]] = {
    lines
      .filter(_.contains(":"))
      .map { line =>
        val parts = line.split(":")
        val device = parts(0).trim
        val outputs = if (parts.length > 1 && parts(1).trim.nonEmpty) {
          parts(1).trim.split("\\s+").toList
        } else {
          List.empty
        }
        device -> outputs
      }
      .toMap
      .withDefaultValue(List.empty)
  }

  /** Count paths from start to end that visit both required nodes (memoized) */
  def countPathsWithBoth(
      graph: Map[String, List[String]],
      start: String,
      end: String,
      required1: String,
      required2: String
  ): Long = {
    // Memoization: (node, hasVisited1, hasVisited2) -> path count
    val memo = mutable.Map[(String, Boolean, Boolean), Long]()

    def dfs(node: String, visited1: Boolean, visited2: Boolean): Long = {
      val v1 = visited1 || (node == required1)
      val v2 = visited2 || (node == required2)

      if (node == end) {
        return if (v1 && v2) 1L else 0L
      }

      val key = (node, v1, v2)
      memo.getOrElseUpdate(key, {
        graph(node).map(child => dfs(child, v1, v2)).sum
      })
    }

    dfs(start, visited1 = false, visited2 = false)
  }

  def solve(filename: String): Long = {
    val lines = Source.fromFile(filename).getLines().toSeq
    val graph = parseGraph(lines)
    countPathsWithBoth(graph, "svr", "out", "dac", "fft")
  }

  def main(args: Array[String]): Unit = {
    val filename = if (args.length > 0) args(0) else "input.txt"
    val result = solve(filename)
    println(s"Part 2: $result")
  }
}

File: 2025/Day01/solution_part2.lua
-- Advent of Code 2025 - Day 1: Secret Entrance (Part 2)
-- Count every click that lands on 0, not just end of rotation

local function solve(filename)
    local dial = 50
    local count = 0

    for line in io.lines(filename) do
        local direction = line:sub(1, 1)
        local distance = tonumber(line:sub(2))

        if direction == "L" then
            if dial == 0 then
                count = count + math.floor(distance / 100)
            elseif distance >= dial then
                count = count + math.floor((distance - dial) / 100) + 1
            end
            dial = (dial - distance) % 100
        elseif direction == "R" then
            count = count + math.floor((dial + distance) / 100)
            dial = (dial + distance) % 100
        end
    end

    return count
end

-- Main
local filename = arg[1] or "input.txt"
print("Password: " .. solve(filename))
